### PR TITLE
qemu_vm: Refactor some network methods which called by hot(un)plug_nic

### PR DESCRIPTION
**Pull request by PaulYuuu**
qemu_monitor:
1. Add 2 qapi methods "netdev_add" and "netdev_del"
2. Migrate QMPMonitor.set_link from self.send_args_cmd to self.cmd

qemu_vm:
1. Refactor to use simple_hot(un)plug to plug/unplug network frontend device
2. Use monitor.netdev_add and monitor_netdev_del to plug/unplug network backend device

ID: 1860240
References: https://github.com/qemu/qemu/tree/master/qapi
Signed-off-by: Yihuang Yu <yihyu@redhat.com>

### CODE REVIEW

**Comment by PaulYuuu** [comment]
Some referances:
https://github.com/qemu/qemu/blob/194f8ca825854abef3aceca1ed7eb5a53b08751f/qapi/opts-visitor.c#L371-L403
```c
/* mimics qemu-option.c::parse_option_bool() */
static bool
opts_type_bool(Visitor *v, const char *name, bool *obj, Error **errp)
{
    OptsVisitor *ov = to_ov(v);
    const QemuOpt *opt;


    opt = lookup_scalar(ov, name, errp);
    if (!opt) {
        return false;
    }


    if (opt->str) {
        if (strcmp(opt->str, "on") == 0 ||
            strcmp(opt->str, "yes") == 0 ||
            strcmp(opt->str, "y") == 0) {
            *obj = true;
        } else if (strcmp(opt->str, "off") == 0 ||
            strcmp(opt->str, "no") == 0 ||
            strcmp(opt->str, "n") == 0) {
            *obj = false;
        } else {
            error_setg(errp, QERR_INVALID_PARAMETER_VALUE, opt->name,
                       "on|yes|y|off|no|n");
            return false;
        }
    } else {
        *obj = true;
    }


    processed(ov, name);
    return true;
}
```

https://github.com/qemu/qemu/blob/194f8ca825854abef3aceca1ed7eb5a53b08751f/qapi/net.json#L234-L250
```json
{ 'struct': 'NetdevTapOptions',
  'data': {
    '*ifname':     'str',
    '*fd':         'str',
    '*fds':        'str',
    '*script':     'str',
    '*downscript': 'str',
    '*br':         'str',
    '*helper':     'str',
    '*sndbuf':     'size',
    '*vnet_hdr':   'bool',
    '*vhost':      'bool',
    '*vhostfd':    'str',
    '*vhostfds':   'str',
    '*vhostforce': 'bool',
    '*queues':     'uint32',
    '*poll-us':    'uint32'} }
```

https://github.com/qemu/qemu/blob/194f8ca825854abef3aceca1ed7eb5a53b08751f/qapi/block-core.json#L173-L180
```json
{ 'struct': 'ImageInfo',
  'data': {'filename': 'str', 'format': 'str', '*dirty-flag': 'bool',
           '*actual-size': 'int', 'virtual-size': 'int',
           '*cluster-size': 'int', '*encrypted': 'bool', '*compressed': 'bool',
           '*backing-filename': 'str', '*full-backing-filename': 'str',
           '*backing-filename-format': 'str', '*snapshots': ['SnapshotInfo'],
           '*backing-image': 'ImageInfo',
           '*format-specific': 'ImageInfoSpecific' } }
```

**Comment by PaulYuuu** [comment]
@luckyh Could you help to review?
@yanglei-rh This one can fix your problem.

**Comment by yanglei-rh** [comment]
@PaulYuuu If so, I will closed 2323.patch and update test result after test finished.

**Comment by PaulYuuu** [comment]
```
2020-07-26 07:39:35: {"execute": "netdev_add", "arguments": {"type": "tap", "id": "idjum5tk", "fd": "96", "vhost": true}, "id": "z
mGlB83y"}
```
```
2020-07-26 23:46:51: {"execute": "set_link", "arguments": {"name": "idZ6tBer", "up": true}, "id": "MsLO1MMZ"}
```

File: virttest/qemu_monitor.py [hunk 1]
```diff
@@ -2222,9 +2222,9 @@ def send_args_cmd(self, cmdlines, timeout=CMD_TIMEOUT, convert=True):
                                 value = int(value)
                         elif re.match("^[0-9]+\.[0-9]*$", value):
                             value = float(value)
-                        elif re.findall("true", value, re.I):
+                        elif value.lower() in ["on", "yes", "y", "true"]:
```
Comment (luckyh) [inline virttest/qemu_monitor.py:2225]: Well, the question is: is it true for all the usages that any string in the list has to be converted to bool values? for example, what if I want `id=y`, is it acceptable?

Comment (PaulYuuu) [reply virttest/qemu_monitor.py:2225]: Nice catch. `y` and `n` have no clear semantics in other scenarios, I think can move it from the list, only use `yes` directly.

**Comment by yanglei-rh** [comment]
RHEL:
(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.vhost_nic.nic_virtio.q35: PASS (127.19 s)
01:23:52 DEBUG| Send command: {'execute': 'netdev_add', 'arguments': {'type': 'tap', 'id': 'id0rNkiF', 'fd': '37', 'vhost': True}, 'id': '6rpmCUxF'}
01:23:52 DEBUG| Send command: {'execute': 'device_add', 'arguments': {'driver': 'virtio-net-pci', 'netdev': 'id0rNkiF', 'mac': '9a:29:39:7b:71:3f', 'id': 'id793egL', 'bus': 'pcie_extra_root_port_0', 'addr': '0x0'}, 'id': 'sQheOybH'}
Windows (exists hotplug/unplug bug):
(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.vhost_nic.nic_virtio.q35: ERROR (186.39 s)
02:28:01 DEBUG| Send command: {'execute': 'netdev_add', 'arguments': {'type': 'tap', 'id': 'idGnzK9j', 'fd': '48', 'vhost': True}, 'id': 'tWxEV3c5'}
02:28:01 DEBUG| Send command: {'execute': 'device_add', 'arguments': {'driver': 'virtio-net-pci', 'netdev': 'idGnzK9j', 'mac': '9a:27:bb:70:0e:3c', 'id': 'idQr7SYH', 'bus': 'pcie_extra_root_port_0', 'addr': '0x0'}, 'id': 'HkuWYTGP'}

**Comment by PaulYuuu** [comment]
@luckyh After our discussion, I updated this PR to handle some network-specific methods, please help to review again, thanks.
@yanglei-rh Can you apply this patch and verify it works well on x86?

**Comment by yanglei-rh** [comment]
LGTM
Acked-by: Lei Yang

Test Pass(except windows bug)
RHEL：
(02/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (240.05 s)
 (03/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.q35: PASS (208.42 s)
 (04/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_repetition.q35: PASS (5196.04 s)
 (05/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_reboot.q35: PASS (170.45 s)
 (06/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_shutdown.after_plug.q35: PASS (96.78 s)
 (07/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_shutdown.after_unplug.q35: PASS (96.26 s)
 (08/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_system_reset.q35: PASS (180.17 s)
 (09/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_e1000e.q35: PASS (99.98 s)
 (10/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.multi_pci.nic_virtio.q35: PASS (254.58 s)
 (11/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.multi_pci.nic_e1000e.q35: PASS (152.37 s)
 (12/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.additional.nic_virtio.q35: PASS (165.08 s)
 (13/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.additional.nic_e1000e.q35: PASS (169.86 s)
(18/20) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.nic_hotplug.vhost_nic.nic_virtio.q35: PASS (171.61 s)
Windows:
 (02/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.q35: PASS (151.25 s)
 (03/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_repetition.q35: ERROR: Shell command failed: 'netsh interface set interface name="C:\\Windows\\system32>" admin=DISABLED'    (status: 1,    output: 'No more data is available.\n\n\n\nC:\\Windows\\system32>\n') (1438.18 s)
 (04/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_reboot.q35: PASS (496.76 s)
 (05/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_shutdown.after_plug.q35: PASS (136.53 s)
 (06/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_shutdown.after_unplug.q35: PASS (154.27 s)
 (07/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_virtio.with_system_reset.q35: PASS (324.15 s)
 (08/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.one_pci.nic_e1000e.q35: PASS (142.96 s)
 (09/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.multi_pci.nic_virtio.q35: PASS (205.55 s)
 (10/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.multi_pci.nic_e1000e.q35: PASS (202.62 s)
 (11/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.additional.nic_virtio.q35: PASS (192.31 s)
 (12/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.additional.nic_e1000e.q35: PASS (149.16 s)
(17/19) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.nic_hotplug.vhost_nic.nic_virtio.q35: ERROR: Timeout expired while waiting for shell command to complete: 'netsh interface set interface name="Ethernet 2" admin=DISABLED'    (output: '') (660.45 s)

**Comment by PaulYuuu** [comment]
Test results with the latest code:

QMP:
```
06:41:35 DEBUG| Context: Start test iteration 1 --> (hotplug_nic) --> Hotplugging nic hotplug_nic1 on vm avocado-vt-vm1
06:41:35 DEBUG| (monitor avocado-vt-vm1.qmpmonitor1) Sending command 'netdev_add'
06:41:35 DEBUG| Send command: {'execute': 'netdev_add', 'arguments': {'fd': '96', 'vhost': True, 'type': 'tap', 'id': 'id8KfzHz'}, 'id': '3j2QKPfa'}
06:41:35 DEBUG| (monitor avocado-vt-vm1.qmpmonitor1) Sending command 'device_add'
06:41:35 DEBUG| Send command: {'execute': 'device_add', 'arguments': OrderedDict([('id', 'id1ZwEgd'), ('driver', 'virtio-net-pci'), ('netdev', 'id8KfzHz'), ('mac', '9a:35:91:d7:dc:27'), ('bus', 'pci.0'), ('addr', '0x5')]), 'id': 'MMb75wDg'}
```

HMP:
```
06:45:14 DEBUG| Context: Start test iteration 1 --> (hotplug_nic) --> Hotplugging nic hotplug_nic1 on vm avocado-vt-vm1
06:45:14 DEBUG| (monitor avocado-vt-vm1.qmpmonitor1) Sending command 'netdev_add type=tap,id=idCPotBp,fd=93,vhost=on'
06:45:14 DEBUG| Send command: netdev_add type=tap,id=idCPotBp,fd=93,vhost=on
06:45:14 DEBUG| (monitor avocado-vt-vm1.qmpmonitor1) Sending command 'device_add virtio-net-pci,id=idR0rAvd,netdev=idCPotBp,mac=9a:93:ee:36:81:c4,bus=pci.0,addr=0x5'
06:45:14 DEBUG| Send command: device_add virtio-net-pci,id=idR0rAvd,netdev=idCPotBp,mac=9a:93:ee:36:81:c4,bus=pci.0,addr=0x5
```

**Comment by PaulYuuu** [comment]
Found a typo and corrected it.
Also, update test results.

QMP:
```
 (1/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.nic_hotplug.migration.after_nichotplug.nic_virtio: PASS (569.51 s)
 (2/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.nic_hotplug.migration.after_nichotunplug.nic_virtio: PASS (333.62 s)
 (3/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.nic_hotplug.vhost_nic.nic_virtio: PASS (196.42 s)
 (4/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.set_link.reboot_shell.status_on: PASS (1200.71 s)
```

HMP:
```
 (1/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.nic_hotplug.migration.after_nichotplug.nic_virtio: PASS (451.98 s)
 (2/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.nic_hotplug.migration.after_nichotunplug.nic_virtio: PASS (285.15 s)
 (3/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.nic_hotplug.vhost_nic.nic_virtio: PASS (157.88 s)
 (4/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.set_link.reboot_shell.status_on: PASS (924.55 s)
```

**Review by luckyh — APPROVED** [review]
LGTM

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**